velox/exec/WriteProtocol.h

```diff
@@ -0,0 +1,173 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/vector/ComplexVector.h"
+
+namespace facebook::velox::connector {
+class ConnectorInsertTableHandle;
+class ConnectorQueryCtx;
+} // namespace facebook::velox::connector
+
+namespace facebook::velox::exec {
+
+class Operator;
+
+enum class CommitStrategy { kNoCommit, kTaskCommit };
+
+class WriteParameters {
+ public:
+  virtual ~WriteParameters() = default;
+};
+
+class HiveWriteParameters : public WriteParameters {
+ public:
+  enum class UpdateMode { kNew, kAppend, kOverwrite };
+
+  HiveWriteParameters(
+      UpdateMode updateMode,
+      const std::string& targetFileName,
+      const std::string& writeFileName,
+      const std::string& targetDirectory,
+      const std::string& writeDirectory)
+      : WriteParameters(),
+        updateMode_(updateMode),
+        targetFileName_(targetFileName),
+        writeFileName_(writeFileName),
+        targetDirectory_(targetDirectory),
+        writeDirectory_(writeDirectory) {}
+
+  ~HiveWriteParameters() {}
+
+  UpdateMode updateMode() const {
+    return updateMode_;
+  }
+
+  std::string encodeUpdateMode() const {
+    switch (updateMode_) {
+      case UpdateMode::kNew:
+        return "NEW";
+      case UpdateMode::kAppend:
+        return "APPEND";
+      case UpdateMode::kOverwrite:
+        return "OVERWRITE";
+      default:
+        VELOX_UNSUPPORTED("Unsupported update mdoe.");
+    }
+  }
+
+  const std::string& targetFileName() const {
+    return targetFileName_;
+  }
+
+  const std::string& writeFileName() const {
+    return writeFileName_;
+  }
+
+  const std::string& targetDirectory() const {
+    return targetDirectory_;
+  }
+
+  const std::string& writeDirectory() const {
+    return writeDirectory_;
+  }
+
+ private:
+  const UpdateMode updateMode_;
+  const std::string targetFileName_;
+  const std::string writeFileName_;
+  const std::string targetDirectory_;
+  const std::string writeDirectory_;
+};
+
+class WriteProtocol {
+ public:
+  virtual ~WriteProtocol(){};
+
+  virtual CommitStrategy getCommitStrategy() const = 0;
+
+  virtual std::string encodeCommitStrategy() const = 0;
+
+  virtual RowVectorPtr commit(velox::exec::Operator& commitOperator) {
+    return nullptr;
+  }
+
+  virtual std::shared_ptr<const WriteParameters> getWriteParameters(
+      const std::shared_ptr<const velox::connector::ConnectorInsertTableHandle>&
+          tableHandle,
+      const velox::connector::ConnectorQueryCtx* FOLLY_NONNULL
+          connectorQueryCtx) const = 0;
+
+  static void registerWriteProtocol(
+      CommitStrategy commitStrategy,
+      const std::function<std::shared_ptr<WriteProtocol>()> writeProtocol);
+
+  static std::shared_ptr<WriteProtocol> newWriteProtocol(
+      CommitStrategy commitStrategy);
+};
+
+class HiveNoCommitWriteProtocol : public WriteProtocol {
```

Comment:
It really feels like WriteProtocols should live with connectors. :)
And exec should be using them (we already do have that, I assume).